File: ch7/points_lines.ml
open Printf

let eps = 1e-9
let near x y = abs_float (x -. y) < eps
let pi = Float.pi

let deg_to_rad d = d *. pi /. 180.
let rad_to_deg r = r *. 180. /. pi

type point = { x: float; y: float }
let point (x, y) = { x; y }

let dist a b = hypot (a.x -. b.x) (a.y -. b.y)

let rotate {x; y} degree =
  let rad = deg_to_rad degree in
  let sin = sin rad in
  let cos = cos rad in
  point(x *. cos -. y *. sin, x *. sin +. y *. cos)

type vector = Vector of float * float

let to_vec a b = Vector(b.x -. a.x, b.y -. a.y)
let scale (Vector(a, b)) s = Vector(a *. s, b *. s)
let translate {x; y} (Vector(a, b)) = point(x +. a, y +. b)
let dot (Vector(a1, b1)) (Vector(a2, b2)) = a1 *. a2 +. b1 *. b2
let norm v = sqrt (dot v v)
let cross (Vector(a1, b1)) (Vector(a2, b2)) = a1 *. b2 -. a2 *. b1

let angle a o b =
  let oa = to_vec o a in
  let ob = to_vec o b in
  acos (dot oa ob /. (norm oa *. norm ob))

let ccw p q r = cross (to_vec p q) (to_vec p r) > eps
let collinear p q r = near (cross (to_vec p q) (to_vec p r)) 0.

module Line : sig
  type line
  val through_points    : point -> point -> line
  val from_point_slope  : point -> float -> line
  val are_parallel      : line -> line -> bool
  val are_same          : line -> line -> bool
  val intersect         : line -> line -> point option
  val closest_point     : line -> point -> point
  val reflection_point  : line -> point -> point
  val to_string         : line -> string
end = struct
  (** ax + by + c = 0 *)
  type line = { a: float; b: float; c: float }

  let through_points p1 p2 =
    if near p1.x p2.x then { a = 1.; b = 0.; c = -. p1.x }
    else
      let a = -. (p1.y -. p2.y) /. (p1.x -. p2.x) in
      { a; b = 1.; c = -. a *. p1.x -. p1.y }

  let from_point_slope {x; y} m =
    let a, b = -. m, 1. in
    { a; b; c = -. ((a *. x) +. (b *. y)) }

  type line2 = { m: float; c: float }

  let through_points2 p1 p2 =
    if near p1.x p2.x then None
    else
      let m = (p1.y -. p2.y) /. (p1.x -. p2.x) in
      let c = p1.y -. m *. p1.x in
      Some { m; c }

  let are_parallel l1 l2 = near l1.a l2.a && near l1.b l2.b

  let are_same l1 l2 = are_parallel l1 l2 && near l1.c l2.c

  let intersect l1 l2 =
    if are_parallel l1 l2 then None
    else
      let x = (l2.b *. l1.c -. l1.b *. l2.c) /. (l2.a *. l1.b -. l1.a *. l2.b) in
      let y =
        if not (near l1.b 0.) then -. (l1.a *. x +. l1.c)
        else -. (l2.a *. x +. l2.c) in
      Some (point (x, y))
  
  let closest_point l ({x; y} as p) =
    if near l.b 0. then point(-. l.c, y) else
    if near l.a 0. then point(x, -. l.c) else
    let perpendicular = from_point_slope p (1. /. l.a) in
    match intersect l perpendicular with
    | Some q -> q
    | None -> assert false
  
  let reflection_point l p =
    let q = closest_point l p in
    let v = to_vec p q in
    translate q v

  let to_string line =
    sprintf "%.2f * x + %.2f * y + %.2f = 0.00" line.a line.b line.c
end

let altitude p a b : float * point =
  let ap = to_vec a p in
  let ab = to_vec a b in
  let u = dot ap ab /. dot ab ab in
  let c = translate a (scale ab u) in
  dist p c, c

let dist_to_segment p a b : float * point =
  let ap = to_vec a p in
  let ab = to_vec a b in
  let u = dot ap ab /. dot ab ab in
  if u < 0. then dist p a, a else
  if u > 1. then dist p b, b else
  altitude p a b

let () =
  (* sorting points demo *)
  let p =
    [ point(2., 2.)
    ; point(4., 3.)
    ; point(2., 4.)
    ; point(6., 6.)
    ; point(2., 6.)
    ; point(6., 5.)] in
  p
    |> List.sort (fun p1 p2 ->
      if near p1.x p2.x then compare p1.y p2.y else compare p1.x p2.x)
    |> List.iter (fun {x; y} -> printf "(%.2f, %.2f)\n" x y);
  
  (* the positions of these 7 points (0-based indexing)
  6   P4      P3  P6
  5           P5
  4   P2
  3       P1
  2   P0
  1
  0 1 2 3 4 5 6 7 8 *)
  let p = p @ [point(8., 6.)] |> Array.of_list in
  printf "Euclidean distance between P[0] and P[5] = %.2f\n" (dist p.(0) p.(5));

  let l1 = Line.through_points p.(0) p.(1) in
  let l2 = Line.through_points p.(0) p.(2) in
  let l3 = Line.through_points p.(2) p.(3) in
  let l4 = Line.through_points p.(2) p.(4) in
  printf "%s\n" (Line.to_string l1);
  printf "%s\n" (Line.to_string l2);
  let bool_to_int = function true -> 1 | false -> 0 in
  printf "l1 & l2 are parallel? %d\n" (Line.are_parallel l1 l2 |> bool_to_int);
  printf "l1 & l3 are parallel? %d\n" (Line.are_parallel l1 l3 |> bool_to_int);
  printf "l1 & l2 are the same? %d\n" (Line.are_same l1  l2 |> bool_to_int);
  printf "l2 & l4 are the same? %d\n" (Line.are_same l2  l4 |> bool_to_int);

  (match Line.intersect l1 l2 with
  | Some {x; y} ->
    printf "l1 & l2 are intersect? 1, at (%.2f, %.2f)\n" x y
  | None -> ());

  let fmt = format_of_string ": (%.2f, %.2f), dist = %.2f\n" in
  let d, {x; y} = altitude p.(0) p.(2) p.(3) in
  printf ("Closest point from P[0] to line         (P[2]-P[3])" ^^ fmt) x y d;
  let ({x; y} as q) = Line.closest_point l3 p.(0) in
  printf ("Closest point from P[0] to line V2      (P[2]-P[3])" ^^ fmt) x y (dist p.(0) q);
  
  [(0, 2, 3); (1, 2, 3); (6, 2, 3)] |> List.iter (fun (i, j, k) ->
    let d, {x; y} = dist_to_segment p.(i) p.(j) p.(k) in
    printf ("Closest point from P[%d] to line SEGMENT (P[%d]-P[%d])" ^^ fmt) i j k x y d
  );
  
  let {x; y} = Line.reflection_point l4 p.(1) in
  printf "Reflection point from P[1] to line      (P[2]-P[4]): (%.2f, %.2f)\n" x y;

  [(0, 4, 3); (0, 2, 1); (4, 3, 6)] |> List.iter (fun (i, j, k) ->
    printf "Angle P[%d]-P[%d]-P[%d] = %.2f\n" i j k (rad_to_deg (angle p.(i) p.(j) p.(k)));
  );

  [(0, 2, 3); (0, 3, 2)] |> List.iter (fun (i, j, k) ->
    printf "P[%d], P[%d], P[%d] form A left turn? %d\n" i j k (ccw p.(i) p.(j) p.(k) |> bool_to_int)
  );

  [(0, 2, 3); (0, 2, 4)] |> List.iter (fun (i, j, k) ->
    printf "P[%d], P[%d], P[%d] are collinear? %d\n" i j k (collinear p.(i) p.(j) p.(k) |> bool_to_int)
  );

  let p, q, r = point(3., 7.), point(11., 13.), point(35., 30.) in
  printf "r is on the %s of line p-r\n" (if ccw p q r then "left" else "right");

  (* the positions of these 6 points
     E<--  4
           3       B D<--
           2   A C
           1
  -4-3-2-1 0 1 2 3 4 5 6
          -1
          -2
   F<--   -3 *)
  let a = point(2., 2.) in
  let b = point(4., 3.) in
  let c = point(3., 2.) in
  let v = to_vec a b in
  
  (* translation *)
  let d = translate c v in
  let point_to_string {x; y} = sprintf "(%.2f, %.2f)" x y in
  printf "D = %s\n" (point_to_string d);
  let e = translate c (scale v 0.5) in
  printf "E = %s\n" (point_to_string e);
  
  (* rotation *)
  printf "B = %s\n" (point_to_string b);
  let f = rotate b 90. in
  printf "F = %s\n" (point_to_string f);
  let g = rotate b 180. in
  printf "G = %s\n" (point_to_string g)

File: ch7/points_lines.py
import math

INF = 10**9
EPS = 1e-9

def DEG_to_RAD(d):
  return d*math.pi/180.0


def RAD_to_DEG(r):
  return r*180.0/math.pi


class point_i:
  def __init__(self, x=0, y=0):
    self.x = x
    self.y = y

class point:
  def __init__(self, x=0, y=0):
    self.x = x
    self.y = y

  def __lt__(self, other):
    return (self.x, self.y) < (other.x, other.y)

  def __eq__(self, other):
    return math.isclose(self.x, other.x) and math.isclose(self.y, other.y)


def dist(p1, p2):
  return math.hypot(p1.x-p2.x, p1.y-p2.y)


def rotate(p, theta):
  rad = DEG_to_RAD(theta)
  x = p.x * math.cos(rad) - p.y * math.sin(rad)
  y = p.x * math.sin(rad) + p.y * math.cos(rad)
  return point(x, y)


class line:
  def __init__(self):
    self.a = 0
    self.b = 0
    self.c = 0

def pointsToLine(p1, p2, l):
  if abs(p1.x - p2.x) < EPS:
    l.a, l.b, l.c = 1.0, 0.0, -p1.x
  else:
    a = -(p1.y - p2.y) / (p1.x - p2.x)
    l.a, l.b, l.c = a, 1.0, -(a * p1.x) - p1.y


class line2:
  def __init__(self):
    self.m = 0
    self.c = 0

def pointsToLine2(p1, p2, l):
  if p1.x == p2.x:
    l.m = INF
    l.c = p1.x
    return 0
  else:
    l.m = (p1.y - p2.y) / (p1.x - p2.x)
    l.c = p1.y - l.m * p1.x
    return 1


def areParallel(l1, l2):
  return math.isclose(l1.a, l2.a) and math.isclose(l1.b, l2.b)

def areSame(l1, l2):
  return areParallel(l1, l2) and math.isclose(l1.c, l2.c)


def areIntersect(l1, l2, p):
  if areParallel(l1, l2):
    return False
  p.x = (l2.b * l1.c - l1.b * l2.c) / (l2.a * l1.b - l1.a * l2.b)
  if not math.isclose(l1.b, 0.0):
    p.y = -(l1.a * p.x + l1.c)
  else:
    p.y = -(l2.a * p.x + l2.c)
  return True


class vec:
  def __init__(self, x=0, y=0):
    self.x = x
    self.y = y

def toVec(a, b):
  return vec(b.x-a.x, b.y-a.y)

def scale(v, s):
  return vec(v.x*s, v.y*s)

def translate(p, v):
  return point(p.x+v.x, p.y+v.y)


def pointSlopeToLine(p, m, l):
  l.a, l.b = -m, 1
  l.c = -((l.a * p.x) + (l.b * p.y))


def closestPoint(l, p, ans):
  if math.isclose(l.b, 0.0):
    ans.x, ans.y = -l.c, p.y
    return
  if math.isclose(l.a, 0.0):
    ans.x, ans.y = p.x, -l.c
    return
  perpendicular = line()
  pointSlopeToLine(p, 1.0/l.a, perpendicular)
  areIntersect(l, perpendicular, ans)


def reflectionPoint(l, p, ans):
  b = point()
  closestPoint(l, p, b)
  v = toVec(p, b)
  ans.x, ans.y = p.x + 2 * v.x, p.y + 2 * v.y


def dot(a, b):
  return a.x * b.x + a.y * b.y

def norm_sq(v):
  return v.x * v.x + v.y * v.y

def angle(a, o, b):
  oa = toVec(o, a)
  ob = toVec(o, b)
  return  math.acos(dot(oa, ob) / math.sqrt(norm_sq(oa) * norm_sq(ob)))


def distToLine(p, a, b, c):
  ap = toVec(a, p)
  ab = toVec(a, b)
  u = dot(ap, ab) / norm_sq(ab)
  s = scale(ab, u)
  c.x, c.y = a.x+s.x, a.y+s.y
  return dist(p, c)


def distToLineSegment(p, a, b, c):
  ap = toVec(a, p)
  ab = toVec(a, b)
  u = dot(ap, ab) / norm_sq(ab)
  if u < 0.0:
    c.x, c.y = a.x, a.y
    return dist(p, a)
  if u > 1.0:
    c.x, c.y = b.x, b.y
    return dist(p, b)
  return distToLine(p, a, b, c)


def cross(a, b):
  return a.x * b.y - a.y * b.x


def ccw(p, q, r):
  return cross(toVec(p, q), toVec(p, r)) > -EPS


def collinear(p, q, r):
  return abs(cross(toVec(p, q), toVec(p, r))) < EPS


if __name__ == '__main__':
  P = [point(2e-9, 0), point(0, 2), point(1e-9, 1)]
  P = sorted(P)
  for pt in P:
    print('%.9lf, %.9lf' % (pt.x, pt.y))

  P1 = point()
  P2 = point()
  P3 = point(0, 1)
  print('%d' % (P1 == P2))
  print('%d' % (P1 == P3))

  P = [point(2, 2), point(4, 3), point(2, 4), point(6, 6), point(2, 6), point(6, 5)]

  P = sorted(P)
  for p in P:
    print('(%.2lf, %.2lf)' % (p.x, p.y))

  P = [point(2, 2), point(4, 3), point(2, 4), point(6, 6), point(2, 6), point(6, 5), point(8, 6)]

  d = dist(P[0], P[5])
  print('Euclidean distance between P[0] and P[5] = %.2lf' % d)

  l1 = line()
  l2 = line()
  l3 = line()
  l4 = line()

  pointsToLine(P[0], P[1], l1)
  print('%.2lf * x + %.2lf * y + %.2lf = 0.00' % (l1.a, l1.b, l1.c))

  pointsToLine(P[0], P[2], l2);
  print('%.2lf * x + %.2lf * y + %.2lf = 0.00' % (l2.a, l2.b, l2.c))


  pointsToLine(P[2], P[3], l3)
  print('l1 & l2 are parallel? %d' % areParallel(l1, l2))
  print('l1 & l3 are parallel? %d' % areParallel(l1, l3))

  pointsToLine(P[2], P[4], l4)
  print('l1 & l2 are the same? %d' % areSame(l1, l2))
  print('l2 & l4 are the same? %d' % areSame(l2, l4))
  
  p12 = point()
  res = areIntersect(l1, l2, p12)
  print('l1 & l2 are intersect? %d, at (%.2lf, %.2lf)' % (res, p12.x, p12.y))

  ans = point()
  d = distToLine(P[0], P[2], P[3], ans)
  print('Closest point from P[0] to line         (P[2]-P[3]): (%.2lf, %.2lf), dist = %.2lf' % (ans.x, ans.y, d))
  closestPoint(l3, P[0], ans)
  print('Closest point from P[0] to line V2      (P[2]-P[3]): (%.2lf, %.2lf), dist = %.2lf' % (ans.x, ans.y, dist(P[0], ans)))

  d = distToLineSegment(P[0], P[2], P[3], ans)
  print('Closest point from P[0] to line SEGMENT (P[2]-P[3]): (%.2lf, %.2lf), dist = %.2lf' % (ans.x, ans.y, d))
  d = distToLineSegment(P[1], P[2], P[3], ans)
  print('Closest point from P[1] to line SEGMENT (P[2]-P[3]): (%.2lf, %.2lf), dist = %.2lf' % (ans.x, ans.y, d))
  d = distToLineSegment(P[6], P[2], P[3], ans)
  print('Closest point from P[6] to line SEGMENT (P[2]-P[3]): (%.2lf, %.2lf), dist = %.2lf' % (ans.x, ans.y, d))

  reflectionPoint(l4, P[1], ans)
  print('Reflection point from P[1] to line      (P[2]-P[4]): (%.2lf, %.2lf)' % (ans.x, ans.y))

  print('Angle P[0]-P[4]-P[3] = %.2lf' % RAD_to_DEG(angle(P[0], P[4], P[3])))
  print('Angle P[0]-P[2]-P[1] = %.2lf' % RAD_to_DEG(angle(P[0], P[2], P[1])))
  print('Angle P[4]-P[3]-P[6] = %.2lf' % RAD_to_DEG(angle(P[4], P[3], P[6])))

  print('P[0], P[2], P[3] form A left turn? %d' % ccw(P[0], P[2], P[3]))
  print('P[0], P[3], P[2] form A left turn? %d' % ccw(P[0], P[3], P[2]))

  print('P[0], P[2], P[3] are collinear? %d' % collinear(P[0], P[2], P[3]))
  print('P[0], P[2], P[4] are collinear? %d' % collinear(P[0], P[2], P[4]))

  p = point(3, 7)
  q = point(11, 13)
  r = point(35, 30)
  print('r is on the %s of line p-q (direction p->q)' % ('left' if ccw(p, q, r) else 'right'))

  A = point(2.0, 2.0)
  B = point(4.0, 3.0)
  v = toVec(A, B)
  C = point(3.0, 2.0)
  D = translate(C, v)
  print('D = (%.2lf, %.2lf)' % (D.x, D.y))
  E = translate(C, scale(v, 0.5))
  print('E = (%.2lf, %.2lf)' % (E.x, E.y))

  print('B = (%.2lf, %.2lf)' % (B.x, B.y))
  F = rotate(B, 90)
  print('F = (%.2lf, %.2lf)' % (F.x, F.y))
  G = rotate(B, 180)
  print('G = (%.2lf, %.2lf)' % (G.x, G.y))
